[Autograd Onboarding Lab] Adding native composite function 'attn', OpInfo, and custom derivative function

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62125
* #62124

Part 3/4 and 4/4 of onboarding lab. This PR will not be merged to master.

NivekT committed:
Adding native composite function 'attn' and OpInfo

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -270,6 +270,8 @@
   dispatch:
     CPU, CUDA: angle_out
 
+- func: attn(Tensor q, Tensor k, Tensor v) -> (Tensor, Tensor)
```
nit: you can give names to the outputs so that the python binding will return a NamedTuple instead of just a tuple.

NivekT replied:
Just added that to a previous commit.

albanD commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1407,6 +1407,20 @@ def sample_inputs_baddbmm(op_info, device, dtype, requires_grad, **kwargs):
+def sample_attn(self, device, dtype, requires_grad=False, **kwargs):
+        samples = []
+        shapes = [(2,2), (3,3), (2,3), (10,10), (100,100)]
+        for shape in shapes:
```
You only test the case where all 3 inputs have the same shape here?

NivekT replied:
Added more test cases with different shapes below.

NivekT committed:
Update on "[Autograd Onboarding Lab] Adding native composite function 'attn' and OpInfo"


Part 3/4 of onboarding lab. This PR will not be merged to master.

[ghstack-poisoned]

NivekT commented:
Part 4/4 of the lab has also been added to this PR.

NivekT committed:
Update on "[Autograd Onboarding Lab] Adding native composite function 'attn', OpInfo, and custom derivative function"


Part 3/4 and 4/4 of onboarding lab. This PR will not be merged to master.

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -3575,6 +3575,47 @@ Tensor gather_with_keepdimed_indices(const Tensor& input, int64_t dim, const Ten
+
+  Tensor grad_out = grad_out_;
+  if (!grad_out.defined()) {
+    grad_out = at::zeros(output.sizes(), at::kDouble);
```
I think you want to optimize this code so that you skip doing computations altogether when the grad_output is not defined.

NivekT replied:
Made changes based on this suggestion.

albanD commented on aten/src/ATen/native/Attention.cpp:
```diff
@@ -0,0 +1,18 @@
+
+//::std::tuple<at::Tensor,at::Tensor> wrapper__attn(const at::Tensor & q, const at::Tensor & v, const at::Tensor & k)
+std::tuple <Tensor, Tensor> attn(const Tensor& q, const Tensor& k, const Tensor& v) {
+    Tensor k_t = k.t();
```
I think you should add some kind of error checking here to make sure that q, k and v have the shape you expect.

NivekT replied:
Made changes based on this suggestion.

albanD replied:
I meant to add tests in the forward, not the backward.
But that's ok for this lab.

albanD reviewed (commented):
Looks mostly ok.
CI needs to be fixed though.

NivekT committed:
Update on "[Autograd Onboarding Lab] Adding native composite function 'attn', OpInfo, and custom derivative function"


Part 3/4 and 4/4 of onboarding lab. This PR will not be merged to master.

[ghstack-poisoned]

NivekT committed:
Update on "[Autograd Onboarding Lab] Adding native composite function 'attn', OpInfo, and custom derivative function"


Part 3/4 and 4/4 of onboarding lab. This PR will not be merged to master.

[ghstack-poisoned]

NivekT committed:
Update on "[Autograd Onboarding Lab] Adding native composite function 'attn', OpInfo, and custom derivative function"


Part 3/4 and 4/4 of onboarding lab. This PR will not be merged to master.

[ghstack-poisoned]

NivekT committed:
Update on "[Autograd Onboarding Lab] Adding native composite function 'attn', OpInfo, and custom derivative function"


Part 3/4 and 4/4 of onboarding lab. This PR will not be merged to master.

[ghstack-poisoned]

NivekT committed:
Update on "[Autograd Onboarding Lab] Adding native composite function 'attn', OpInfo, and custom derivative function"


Part 3/4 and 4/4 of onboarding lab. This PR will not be merged to master.

[ghstack-poisoned]

NivekT committed:
Update on "[Autograd Onboarding Lab] Adding native composite function 'attn', OpInfo, and custom derivative function"


Part 3/4 and 4/4 of onboarding lab. This PR will not be merged to master.

[ghstack-poisoned]

NivekT committed:
Update on "[Autograd Onboarding Lab] Adding native composite function 'attn', OpInfo, and custom derivative function"


Part 3/4 and 4/4 of onboarding lab. This PR will not be merged to master.

[ghstack-poisoned]

albanD reviewed (approved):
Looks good to me! :)

NivekT commented:
Closing this because this was done (the remaining CI failures are because of merging with master branch, confirmed with @albanD).
